tests/test_messaging.py

```diff
@@ -1316,6 +1335,384 @@ def test_send_fcm_error_code(self, status):
+            'error': {
+                'status': 'INVALID_ARGUMENT',
+                'message': 'test error',
+                'details': [
+                    {
+                        '@type': 'type.googleapis.com/google.firebase.fcm.v1.FcmError',
+                        'errorCode': 'UNREGISTERED',
+                    },
+                ],
+            }
+        }))
+        _ = self._instrument_batch_messaging_service(payload=payload)
+        msg = messaging.Message(topic='foo')
+        batch_response = messaging.send_all([msg])
+        assert batch_response.success_count is 0
+        assert batch_response.failure_count is 1
+        assert len(batch_response.responses) == 1
+        assert not any([r.message_id for r in batch_response.responses])
+        assert not all([r.success for r in batch_response.responses])
+        exception = batch_response.responses[0].exception
+        assert str(exception) == 'test error'
+        assert str(exception.code) == 'registration-token-not-registered'
+
+    @pytest.mark.parametrize('status', HTTP_ERRORS)
+    def test_send_all_batch_error(self, status):
+        _ = self._instrument_batch_messaging_service(status=status, payload='{}')
+        msg = messaging.Message(topic='foo')
+        with pytest.raises(messaging.ApiCallError) as excinfo:
+            messaging.send_all([msg])
+        expected = 'Unexpected HTTP response with status: {0}; body: {{}}'.format(status)
+        assert str(excinfo.value) == expected
+        assert str(excinfo.value.code) == messaging._MessagingService.UNKNOWN_ERROR
+
+    @pytest.mark.parametrize('status', HTTP_ERRORS)
+    def test_send_all_batch_detailed_error(self, status):
+        payload = json.dumps({
+            'error': {
+                'status': 'INVALID_ARGUMENT',
+                'message': 'test error'
+            }
+        })
+        _ = self._instrument_batch_messaging_service(status=status, payload=payload)
+        msg = messaging.Message(topic='foo')
+        with pytest.raises(messaging.ApiCallError) as excinfo:
+            messaging.send_all([msg])
+        assert str(excinfo.value) == 'test error'
+        assert str(excinfo.value.code) == 'invalid-argument'
+
+    @pytest.mark.parametrize('status', HTTP_ERRORS)
+    def test_send_all_batch_canonical_error_code(self, status):
+        payload = json.dumps({
+            'error': {
+                'status': 'NOT_FOUND',
+                'message': 'test error'
+            }
+        })
+        _ = self._instrument_batch_messaging_service(status=status, payload=payload)
+        msg = messaging.Message(topic='foo')
+        with pytest.raises(messaging.ApiCallError) as excinfo:
+            messaging.send_all([msg])
+        assert str(excinfo.value) == 'test error'
+        assert str(excinfo.value.code) == 'registration-token-not-registered'
+
+    @pytest.mark.parametrize('status', HTTP_ERRORS)
+    def test_send_all_batch_fcm_error_code(self, status):
```

Comment:
I think this is a good start. But I'd recommend making a couple of changes:
1. Change all tests to have at least 2 messages in a batch.
2. Include some tests to check what happens when some messages in a batch are delivered successfully while some fail.